File: rdos-validation-tools/rdo-validation-tool-scala/src/main/scala/com/rms/rdo/validation/cdl/CDLValidator.scala
package com.rms.rdo.validation.cdl

import java.io.File

import com.rms.cdl.parser.validator.ContractValidator
import com.rms.rdo.validation.model.CDLValidationError

import scala.io.Source
import scala.util.Try

object CDLValidator {

    def validateCDLFile(file: File): Option[CDLValidationError] = {
        val fileContent = Try(Source.fromFile(file.getPath)).toOption.map(_.getLines.mkString)

        fileContent.flatMap(content =>
                                try {
                                    ContractValidator.validate(content)
                                    None
                                }
                                catch {
                                    case e: Exception => Some(CDLValidationError(file.getName, e.getMessage))
                                })
    }
}

File: rdos-validation-tools/rdo-validation-tool-java/src/main/java/com/rms/rdo/validation/cdl/CDLValidator.java
package com.rms.rdo.validation.cdl;

import com.rms.cdl.parser.validator.ContractValidator;
import com.rms.rdo.validation.constants.Constants;
import com.rms.rdo.validation.model.cdl.CDLValidationError;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.Optional;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class CDLValidator {

    public static Optional<CDLValidationError> validateCDLFile(File file) {
        Optional<String> fileContent = readFile(file);

        return fileContent.flatMap(content -> {
            try {
                ContractValidator.validate(content);
                return Optional.empty();
            } catch (Exception ex) {
                return Optional.of(CDLValidationError.builder()
                        .file(file.getName()).errorMessage(ex.getMessage()).build());
            }
        });
    }

    private static Optional<String> readFile(File file) {
        try (Stream<String> stream = Files.lines(Paths.get(file.getPath()))) {
            return Optional.ofNullable(stream.collect(Collectors.joining(Constants.EMPTY_STRING)));
        } catch (IOException e) {
            return Optional.empty();
        }
    }
}